api/src/main/java/io/opencensus/trace/TraceState.java

```diff
@@ -0,0 +1,173 @@
+ * limitations under the License.
+ */
+
+package io.opencensus.trace;
+
+import io.opencensus.internal.Utils;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * Carries tracing-system specific context in a list of key-value pairs. TraceState allows different
+ * vendors propagate additional information and inter-operate with their legacy Id formats.
+ *
+ * <p>Implementation is optimized for a small list of key-value pairs.
+ *
+ * <p>Key is opaque string up to 256 characters printable. It MUST begin with a lowercase letter,
+ * and can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
+ * forward slashes /.
+ *
+ * <p>Value is opaque string up to 256 characters printable ASCII RFC0020 characters (i.e., the
+ * range 0x20 to 0x7E) except comma , and =.
+ *
+ * @since 0.16
+ */
+@Immutable
+public final class TraceState {
+  private static final int KEY_MAX_SIZE = 256;
+  private static final int VALUE_MAX_SIZE = 256;
+  private static final int MAX_KEY_VALUE_PAIRS = 128;
+
+  // Immutable list of key-value pairs.
+  private final List<Entry> entries;
+
+  public static final TraceState EMPTY =
+      new TraceState(Collections.unmodifiableList(new ArrayList<Entry>()));
+
+  private TraceState(List<Entry> entries) {
+    this.entries = entries;
+  }
+
+  /**
+   * Returns the value to which the specified key is mapped, or null if this map contains no mapping
+   * for the key.
+   *
+   * @param key with which the specified value is to be associated
+   * @return the value to which the specified key is mapped, or null if this map contains no mapping
+   *     for the key.
+   */
+  @Nullable
+  public String get(String key) {
+    for (Entry entry : entries) {
+      if (entry.key.equals(key)) {
+        return entry.value;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Returns a {@link List} view of the mappings contained in this {@code TraceState}.
+   *
+   * @return a {@link List} view of the mappings contained in this {@code TraceState}.
+   */
+  public List<Entry> entryList() {
+    return entries;
+  }
+
+  /**
+   * Creates a TraceState by appending the extra entries to the {@code parent} in front of the
+   * key-value pairs list and removing duplicate entries.
+   *
+   * @param parent the parent {@code TraceState}.
+   * @param extraEntries the list of extra entries.
+   * @return a TraceState by appending the extra entries to the {@code parent}.
+   */
+  public static TraceState create(TraceState parent, List<Entry> extraEntries) {
+    Utils.checkNotNull(parent, "parent");
+    Utils.checkNotNull(extraEntries, "extraEntries");
+    if (extraEntries.size() == 0) {
+      return parent;
+    }
+
+    for (Entry entry : extraEntries) {
+      Utils.checkArgument(validateKey(entry.key), "Invalid key " + entry.key);
```

Comment:
so this is a typical practice to throw for invalid arguments? Need to be careful with this class